core/src/main/java/io/grpc/NameResolverRegistry.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2015, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc;
+
+import java.net.URI;
+import java.util.LinkedList;
+
+/**
+ * A registry that holds various {@link NameResolver.Factory}s and dispatches target URI to the
+ * first one that can handle it.
+ */
+@ExperimentalApi
+public final class NameResolverRegistry extends NameResolver.Factory {
+  private static final NameResolverRegistry defaultRegistry = new NameResolverRegistry();
+
+  private final LinkedList<NameResolver.Factory> registry = new LinkedList<NameResolver.Factory>();
+
+  private NameResolverRegistry() {
+    // To prevent instantiation
+  }
+
+  static {
+    defaultRegistry.register(DnsNameResolverFactory.getInstance());
+  }
+
+  public static NameResolverRegistry getDefaultRegistry() {
+    return defaultRegistry;
+  }
+
+  /**
+   * Registers a {@link NameResolver.Factory}.
+   */
+  public synchronized void register(NameResolver.Factory factory) {
```

Comment:
Using a map could also be transformed into an SPI. My concern about this is that order matters and you won't notice if a scheme is double-registered. It also has no avenues for optimization because the registry is basically knowledge-less.

We can leave it as-is for now, but we'd need to decide on things before it goes stable.